### PR TITLE
Add a linspace class method

### PR DESCRIPTION
**Pull request by jeanconn**
## Description

Add linspace CxoTime class method

### Unit tests

- [x] Mac
```
(ska3-flight-latest) flame:cxotime jean$ git rev-parse HEAD
566ee5f6a81f343cf0e490c314804ed7365fc5c3
(ska3-flight-latest) flame:cxotime jean$ pytest
========================================================= test session starts =========================================================
platform darwin -- Python 3.11.8, pytest-8.0.2, pluggy-1.4.0
PyQt5 5.15.9 -- Qt runtime 5.15.8 -- Qt compiled 5.15.8
rootdir: /Users/jean/git
configfile: pytest.ini
plugins: astropy-0.11.0, qt-4.4.0, cov-5.0.0, timeout-2.2.0, remotedata-0.4.1, anyio-4.3.0, filter-subpackage-0.2.0, doctestplus-1.2.1, astropy-header-0.2.2, hypothesis-6.112.0, arraydiff-0.6.1, mock-3.14.0
collected 250 items                                                                                                                   

cxotime/tests/test_cxotime.py ................................................................................................. [ 38%]
............................................................................................................................... [ 89%]
...........                                                                                                                     [ 94%]
cxotime/tests/test_cxotime_linspace.py ...............                                                                          [100%]

========================================================= 250 passed in 1.42
```

Independent check of unit tests by [REVIEWER NAME]
- [ ] [PLATFORM]:

### Functional tests

No functional testing.

### CODE REVIEW

File: cxotime/tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import astropy.units as u
+import numpy as np
+import pytest
+
+from cxotime import CxoTime, get_range_in_chunks
+
+
+@pytest.mark.parametrize(
+    "start, stop, dt_max, expected_len, expected_values",
+    [
```
Comment (jeanconn) [inline cxotime/tests/test_utils.py:10]: The ruff formatting made this ugly enough that maybe this should be de-parameterized.

File: cxotime/__init__.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 from .convert import *  # noqa: F401, F403
 from .cxotime import *  # noqa: F401, F403
+from .utils import get_range_in_chunks
```
Comment (taldcroft) [inline cxotime/__init__.py:8]: Generally you should follow the same pattern `from .utils import *` in the top-level `__init__` and then define the imports with `__all__`. But I'm recommending to do away with `utils` so this would be OBE.

File: cxotime/utils.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+import astropy.units as u
+import numpy as np
+
+from cxotime import CxoTime, CxoTimeLike
+
+
+def get_range_in_chunks(start: CxoTimeLike, stop: CxoTimeLike, dt_max: u.Quantity):
```
Comment (taldcroft) [inline cxotime/utils.py:8]: Maybe this would be best as a `CxoTime` class method `linspace`. This would accept `start, stop, num: int | None, step_max: u.Quantity | None`. This borrows from numpy `linspace` and also `step` from `arange`. So a call would look like:
```
from cxotime import CxoTime
times = CxoTime.linspace(start, stop, step_max=1 * u.hr)
```

File: cxotime/cxotime.py [hunk 4]
```diff
@@ -174,6 +175,65 @@ def __init__(self, *args, **kwargs):
...
+
+        if num is not None and step_max is not None:
+            raise ValueError("Only one of num or step_max can be defined")
+
+        if num is not None:
+            if num <= 0:
+                raise ValueError("num must be positive nonzero")
+            times = np.linspace(start, stop, num + 1)
+        else:
+            # Require that step_max is a positive nonzero quantity
```
Comment (taldcroft) [inline cxotime/cxotime.py:223]: This `else` block only needs to compute the correct `num`, and then that can be passed to a common `times = np.linspace(start, stop, num + 1)` outside of the if / else.

File: cxotime/tests/test_cxotime_linspace.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+            np.isclose((result[1:] - result[:-1]).sec, dt_max.to(u.second).value)
+        )
+
+    # Confirm that the time range is covered
+    assert result[0] == start
+    assert result[-1] == stop
+
+    # And confirm that the result is as expected
+    assert len(result) == expected_len
+    assert all(a == CxoTime(b) for a, b in zip(result, expected_values, strict=False))
```
Comment (taldcroft) [inline cxotime/tests/test_cxotime_linspace.py:44]: Use this everywhere instead of `zip`, where you don't need to pass `expected_len` since that is defined uniquely by `expected_values`.
```suggestion
assert np.all(result == CxoTime(expected_values))
```

File: cxotime/tests/test_cxotime_linspace.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            "2000:001",
+            "2000:005",
+            24 * u.hour,
+            5,
+            ["2000:001", "2000:002", "2000:003", "2000:004", "2000:005"],
+        ),
+        (
+            "2000:001",
+            "2000:005",
+            2880 * u.minute,
```
Comment (taldcroft) [inline cxotime/tests/test_cxotime_linspace.py:22]: The first two tests are exact integral divisors of 5 days, so they are basically the same test case. 

This test needs coverage of the case of a value that is not an integral divisor. There is no real need to change the units in these tests, you can trust that 1 * u.hour == 60 * u.minute == 3600 * u.s. (I had to open a terminal to see what `2880 / 60` was).

File: cxotime/tests/test_cxotime_linspace.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+            ["2000:001", "2000:003", "2000:005"],
+        ),
+        ("2000:001", "2000:005", 10 * 86400 * u.second, 2, ["2000:001", "2000:005"]),
+    ],
+)
+def test_linspace_step(start, stop, dt_max, expected_len, expected_values):
+    result = CxoTime.linspace(start, stop, step_max=dt_max)
+
+    # Confirm that the time intervals are uniform if there is any interval
+    if len(result) > 2:
```
Comment (taldcroft) [inline cxotime/tests/test_cxotime_linspace.py:33]: The function *always* returns at least one interval and so has length >= 2. The indexing below would work for length==2, but it isn't quite the right test anyway.

The things to test are:
- All the interval durations are the same, regardless of the value.
- That one interval duration is `<= min(dt_max, stop - start)`.

File: cxotime/tests/test_cxotime_linspace.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    stop = CxoTime("2000:001")
+    dt_max = 24 * u.hour
+    result = CxoTime.linspace(start, stop, step_max=dt_max)
+    assert len(result) == 5
+    expected_values = ["2000:005", "2000:004", "2000:003", "2000:002", "2000:001"]
+    assert all(a == CxoTime(b) for a, b in zip(result, expected_values, strict=False))
+
+
+# Add a test that shows that the time range is covered even if the time range is less than dt_max
+def test_linspace_big_step():
```
Comment (taldcroft) [inline cxotime/tests/test_cxotime_linspace.py:59]: You already have a test of this.

File: cxotime/tests/test_cxotime_linspace.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    start = CxoTime("2000:005")
+    stop = CxoTime("2000:001")
+    dt_max = 24 * u.hour
+    result = CxoTime.linspace(start, stop, step_max=dt_max)
+    assert len(result) == 5
+    expected_values = ["2000:005", "2000:004", "2000:003", "2000:002", "2000:001"]
+    assert all(a == CxoTime(b) for a, b in zip(result, expected_values, strict=False))
+
+
+# Add a test that shows that the time range is covered even if the time range is less than dt_max
```
Comment (taldcroft) [inline cxotime/tests/test_cxotime_linspace.py:58]: All these pre-function comments should be test function docstrings like:
```
"""Time range is covered even if the time range is less than dt_max"""
```

File: cxotime/tests/test_cxotime_linspace.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+
+# Add a test that shows we get an error if dt_max is zero
+def test_linspace_zero_step():
+    start = CxoTime("2020:001")
+    stop = CxoTime("2020:005")
+    dt_max = 0 * u.day
+    with pytest.raises(ValueError) as excinfo:
+        CxoTime.linspace(start, stop, step_max=dt_max)
+    assert "step_max must be positive nonzero" in str(excinfo.value)
```
Comment (taldcroft) [inline cxotime/tests/test_cxotime_linspace.py:77]: Use the `match` parameter of `pytest.raises`
```suggestion
    with pytest.raises(ValueError, match="step_max must be positive nonzero"):
        CxoTime.linspace(start, stop, step_max=dt_max)
```

**Review by taldcroft — CHANGES_REQUESTED** [review]

File: cxotime/cxotime.py [hunk 11]
```diff
@@ -174,6 +175,65 @@ def __init__(self, *args, **kwargs):
...
+            if step_max <= 0 * u.s:
+                raise ValueError("step_max must be positive nonzero")
+
+            # Let this work if start > stop, but flip the sign of dt_max
+            if start > stop:
+                step_max = -step_max
+
+            # Calculate chunks to cover time range, handling edge case of start == stop
+            n_chunk = max(np.ceil(float((stop - start) / step_max)), 1)
```
[the remark this answers is not in the document]
Comment (jeanconn) [reply cxotime/cxotime.py]: That's fine, though I did need to cast to integer.

File: cxotime/tests/test_cxotime_linspace.py (hunk 7 above)
Comment (jeanconn) [reply cxotime/tests/test_cxotime_linspace.py]: I needed some abs() on that interval duration - you can let me know if that's what you had in mind for the different positive/negative cases.

File: cxotime/cxotime.py [hunk 12]
```diff
@@ -174,6 +175,65 @@ def __init__(self, *args, **kwargs):
...
+        Returns
+        -------
+        CxoTime
+            CxoTime with time bin edges for each chunk.
+        """
+        start = CxoTime(start)
+        stop = CxoTime(stop)
+
+        if num is not None and step_max is not None:
+            raise ValueError("Only one of num or step_max can be defined")
```
[the remark this answers is not in the document]
Comment (jeanconn) [reply cxotime/cxotime.py:213]: Oh, I think you just meant '=='.

[the remark this answers is not in the document]
Comment (taldcroft) [reply cxotime/cxotime.py]: Yup, sorry. `==`. This is the `xor` function that Python does not provide natively.
